synapse/storage/databases/main/pusher.py

```diff
@@ -373,3 +373,44 @@ def delete_pusher_txn(txn, stream_id):
             await self.db_pool.runInteraction(
                 "delete_pusher", delete_pusher_txn, stream_id
             )
+
+    async def delete_all_pushers_for_user(self, user_id: str) -> None:
+        """Delete all pushers associated with an account."""
+
+        # We want to generate a row in `deleted_pushers` for each pusher we're
+        # deleting, so we fetch the list now so we can generate the appropriate
+        # number of stream IDs.
+        #
+        # Note: technically there could be a race here between adding/deleting
+        # pushers, but a) the worst case if we don't stop a pusher until the
+        # next restart and b) this is only called when we're deactivating an
+        # account.
+        pushers = list(await self.get_pushers_by_user_id(user_id))
+
+        def delete_pushers_txn(txn, stream_ids):
+            self._invalidate_cache_and_stream(  # type: ignore
+                txn, self.get_if_user_has_pusher, (user_id,)
+            )
+
+            self.db_pool.simple_delete_txn(
+                txn,
+                table="pushers",
+                keyvalues={"user_name": user_id},
+            )
+
+            for stream_id, pusher in zip(stream_ids, pushers):
+                self.db_pool.simple_insert_txn(
```

Comment:
Looks like this could be `simple_insert_many_txn`, but not sure if that's an optimization or not. (Is there any reason *not* to do that?)

Reply:
Good point, `simple_insert_many` is quicker